sdk/servicebus/azure-servicebus/azure/servicebus/_common/auto_lock_renewer.py

```diff
@@ -0,0 +1,128 @@
+# ------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# -------------------------------------------------------------------------
+
+import sys
+import datetime
+import logging
+import threading
+import time
+import functools
+from concurrent.futures import ThreadPoolExecutor
+
+from .._servicebus_session import ServiceBusSession
+from ..exceptions import AutoLockRenewFailed, AutoLockRenewTimeout, ServiceBusError
+from .utils import renewable_start_time, utc_now
+
+_log = logging.getLogger(__name__)
+
+class AutoLockRenew(object):
+    """Auto renew locks for messages and sessions using a background thread pool.
+
+    :param executor: A user-specified thread pool. This cannot be combined with
+     setting `max_workers`.
+    :type executor: ~concurrent.futures.ThreadPoolExecutor
+    :param max_workers: Specify the maximum workers in the thread pool. If not
+     specified the number used will be derived from the core count of the environment.
+     This cannot be combined with `executor`.
+    :type max_workers: int
+
+    .. admonition:: Example:
+
+        .. literalinclude:: ../samples/sync_samples/sample_code_servicebus.py
+            :start-after: [START auto_lock_renew_message_sync]
+            :end-before: [END auto_lock_renew_message_sync]
+            :language: python
+            :dedent: 4
+            :caption: Automatically renew a message lock
+
+        .. literalinclude:: ../samples/sync_samples/sample_code_servicebus.py
+            :start-after: [START auto_lock_renew_session_sync]
+            :end-before: [END auto_lock_renew_session_sync]
+            :language: python
+            :dedent: 4
+            :caption: Automatically renew a session lock
+
+    """
+
+    def __init__(self, executor=None, max_workers=None):
+        self.executor = executor or ThreadPoolExecutor(max_workers=max_workers)
```

Comment:
Same comment as the loop, I think this should be internal `self._executor` - unless there's a reason someone could need to interact with it?